# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [23]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [24]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    eos = source_vocab_to_int['<EOS>']
    source = [[source_vocab_to_int[word] for word in sentence.split()] for sentence in source_text.split('\n')]
    target = [[target_vocab_to_int[word] for word in sentence.split()] + [eos] for sentence in target_text.split('\n')]
    return source, target

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [25]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [26]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [28]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    inputs = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='target')
    learning_rate = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None], name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, [None], name='source_sequence_length')
    return (inputs, targets, learning_rate, keep_prob, 
            target_sequence_length, max_target_sequence_length, source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [29]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    go = source_vocab_to_int['<GO>']
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    return tf.concat([tf.fill([batch_size, 1], go), ending], 1)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [31]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # Encoder embedding
    encoder_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)

    # RNN cell
    def make_cell(rnn_size):
        encoder_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return tf.contrib.rnn.DropoutWrapper(encoder_cell, output_keep_prob=keep_prob)

    encoder_cell = tf.contrib.rnn.MultiRNNCell([ make_cell(rnn_size) for _ in range(num_layers)])
    
    encoder_output, encoder_state = tf.nn.dynamic_rnn(encoder_cell, encoder_embed_input, 
                                              sequence_length=source_sequence_length, dtype=tf.float32)
    
    return encoder_output, encoder_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [32]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    drop = tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
    helper = tf.contrib.seq2seq.TrainingHelper(dec_embed_input, target_sequence_length)
    decoder = tf.contrib.seq2seq.BasicDecoder(drop, helper, encoder_state, output_layer)
    training_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(decoder, 
                                                                   impute_finished=True, 
                                                                   maximum_iterations=max_summary_length)
    return training_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [33]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    drop = tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    infer_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)
    # Basic decoder
    infer_decoder = tf.contrib.seq2seq.BasicDecoder(drop, infer_helper, encoder_state, output_layer)
    infer_decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(infer_decoder,
                                                            impute_finished=True,
                                                            maximum_iterations=max_target_sequence_length)
    return infer_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # 1. Decoder Embedding
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    # 3. Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    # 4. Set up a training decoder and an inference decoder
    # Training Decoder
    with tf.variable_scope("decode"):
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                                                       target_sequence_length, max_target_sequence_length,
                                                        output_layer, keep_prob)
    
    # 5. Inference Decoder
    # Reuses the same parameters trained by the training process
    with tf.variable_scope("decode", reuse=True):
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, 
                             target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], 
                             max_target_sequence_length, target_vocab_size, output_layer, batch_size, keep_prob)


    return training_decoder_output, inference_decoder_output
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Pass the input data through the encoder. ignore the encoder output, but use the state
    _, enc_state = encoding_layer(input_data, 
                                  rnn_size, 
                                  num_layers, 
                                  source_sequence_length = source_sequence_length,
                                  source_vocab_size = source_vocab_size, 
                                  encoding_embedding_size = enc_embedding_size,
                                  keep_prob = keep_prob)
    
    # Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    # Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, 
                                                                       enc_state,
                                                                       target_sequence_length,
                                                                       max_target_sentence_length,
                                                                       rnn_size, 
                                                                       num_layers, 
                                                                       target_vocab_to_int,
                                                                       target_vocab_size,
                                                                       batch_size,
                                                                       keep_prob,
                                                                       dec_embedding_size) 
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [14]:
# Number of Epochs
epochs = 60
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 50
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.7
display_step = 20

### Build the Graph
Build the graph using the neural network you implemented.

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [16]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [17]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   20/1077 - Train Accuracy: 0.2316, Validation Accuracy: 0.3050, Loss: 4.6358
Epoch   0 Batch   40/1077 - Train Accuracy: 0.2316, Validation Accuracy: 0.3050, Loss: 3.7600
Epoch   0 Batch   60/1077 - Train Accuracy: 0.2958, Validation Accuracy: 0.3356, Loss: 3.4023
Epoch   0 Batch   80/1077 - Train Accuracy: 0.2707, Validation Accuracy: 0.3363, Loss: 3.3756
Epoch   0 Batch  100/1077 - Train Accuracy: 0.2617, Validation Accuracy: 0.3363, Loss: 3.3328
Epoch   0 Batch  120/1077 - Train Accuracy: 0.2875, Validation Accuracy: 0.3505, Loss: 3.2398
Epoch   0 Batch  140/1077 - Train Accuracy: 0.2603, Validation Accuracy: 0.3590, Loss: 3.3185
Epoch   0 Batch  160/1077 - Train Accuracy: 0.3059, Validation Accuracy: 0.3633, Loss: 3.0719
Epoch   0 Batch  180/1077 - Train Accuracy: 0.3449, Validation Accuracy: 0.4031, Loss: 3.0272
Epoch   0 Batch  200/1077 - Train Accuracy: 0.3570, Validation Accuracy: 0.4158, Loss: 2.9599
Epoch   0 Batch  220/1077 - Train Accuracy: 0.3372, Validati

Epoch   1 Batch  720/1077 - Train Accuracy: 0.4597, Validation Accuracy: 0.4911, Loss: 1.0335
Epoch   1 Batch  740/1077 - Train Accuracy: 0.5074, Validation Accuracy: 0.5192, Loss: 0.9431
Epoch   1 Batch  760/1077 - Train Accuracy: 0.5035, Validation Accuracy: 0.5202, Loss: 0.9648
Epoch   1 Batch  780/1077 - Train Accuracy: 0.4734, Validation Accuracy: 0.5227, Loss: 0.9610
Epoch   1 Batch  800/1077 - Train Accuracy: 0.4496, Validation Accuracy: 0.5312, Loss: 0.9534
Epoch   1 Batch  820/1077 - Train Accuracy: 0.4500, Validation Accuracy: 0.5320, Loss: 0.9660
Epoch   1 Batch  840/1077 - Train Accuracy: 0.5027, Validation Accuracy: 0.5167, Loss: 0.9169
Epoch   1 Batch  860/1077 - Train Accuracy: 0.4985, Validation Accuracy: 0.5213, Loss: 0.9055
Epoch   1 Batch  880/1077 - Train Accuracy: 0.5184, Validation Accuracy: 0.5291, Loss: 0.9061
Epoch   1 Batch  900/1077 - Train Accuracy: 0.4891, Validation Accuracy: 0.5138, Loss: 0.9163
Epoch   1 Batch  920/1077 - Train Accuracy: 0.4777, Validati

Epoch   3 Batch  360/1077 - Train Accuracy: 0.5637, Validation Accuracy: 0.5774, Loss: 0.6076
Epoch   3 Batch  380/1077 - Train Accuracy: 0.5777, Validation Accuracy: 0.5973, Loss: 0.5922
Epoch   3 Batch  400/1077 - Train Accuracy: 0.5965, Validation Accuracy: 0.5717, Loss: 0.6108
Epoch   3 Batch  420/1077 - Train Accuracy: 0.5816, Validation Accuracy: 0.5955, Loss: 0.5824
Epoch   3 Batch  440/1077 - Train Accuracy: 0.5871, Validation Accuracy: 0.5838, Loss: 0.6074
Epoch   3 Batch  460/1077 - Train Accuracy: 0.5668, Validation Accuracy: 0.5980, Loss: 0.6097
Epoch   3 Batch  480/1077 - Train Accuracy: 0.5814, Validation Accuracy: 0.5962, Loss: 0.6079
Epoch   3 Batch  500/1077 - Train Accuracy: 0.5613, Validation Accuracy: 0.5863, Loss: 0.5962
Epoch   3 Batch  520/1077 - Train Accuracy: 0.6376, Validation Accuracy: 0.5920, Loss: 0.5505
Epoch   3 Batch  540/1077 - Train Accuracy: 0.5941, Validation Accuracy: 0.5906, Loss: 0.5480
Epoch   3 Batch  560/1077 - Train Accuracy: 0.5707, Validati

Epoch   4 Batch 1060/1077 - Train Accuracy: 0.6250, Validation Accuracy: 0.6364, Loss: 0.4460
Epoch   5 Batch   20/1077 - Train Accuracy: 0.6145, Validation Accuracy: 0.6300, Loss: 0.4087
Epoch   5 Batch   40/1077 - Train Accuracy: 0.6156, Validation Accuracy: 0.6452, Loss: 0.4248
Epoch   5 Batch   60/1077 - Train Accuracy: 0.6269, Validation Accuracy: 0.6364, Loss: 0.4061
Epoch   5 Batch   80/1077 - Train Accuracy: 0.6094, Validation Accuracy: 0.6396, Loss: 0.4236
Epoch   5 Batch  100/1077 - Train Accuracy: 0.6219, Validation Accuracy: 0.6388, Loss: 0.4226
Epoch   5 Batch  120/1077 - Train Accuracy: 0.6535, Validation Accuracy: 0.6339, Loss: 0.4255
Epoch   5 Batch  140/1077 - Train Accuracy: 0.6172, Validation Accuracy: 0.6413, Loss: 0.4272
Epoch   5 Batch  160/1077 - Train Accuracy: 0.6551, Validation Accuracy: 0.6431, Loss: 0.4012
Epoch   5 Batch  180/1077 - Train Accuracy: 0.6336, Validation Accuracy: 0.6548, Loss: 0.4015
Epoch   5 Batch  200/1077 - Train Accuracy: 0.6473, Validati

Epoch   6 Batch  700/1077 - Train Accuracy: 0.6602, Validation Accuracy: 0.6520, Loss: 0.3101
Epoch   6 Batch  720/1077 - Train Accuracy: 0.6834, Validation Accuracy: 0.6474, Loss: 0.3519
Epoch   6 Batch  740/1077 - Train Accuracy: 0.7109, Validation Accuracy: 0.6541, Loss: 0.3081
Epoch   6 Batch  760/1077 - Train Accuracy: 0.6813, Validation Accuracy: 0.6829, Loss: 0.3272
Epoch   6 Batch  780/1077 - Train Accuracy: 0.6910, Validation Accuracy: 0.6797, Loss: 0.3211
Epoch   6 Batch  800/1077 - Train Accuracy: 0.7133, Validation Accuracy: 0.6662, Loss: 0.3094
Epoch   6 Batch  820/1077 - Train Accuracy: 0.6605, Validation Accuracy: 0.6857, Loss: 0.3334
Epoch   6 Batch  840/1077 - Train Accuracy: 0.7344, Validation Accuracy: 0.6932, Loss: 0.3007
Epoch   6 Batch  860/1077 - Train Accuracy: 0.6849, Validation Accuracy: 0.6591, Loss: 0.3051
Epoch   6 Batch  880/1077 - Train Accuracy: 0.7047, Validation Accuracy: 0.6751, Loss: 0.3031
Epoch   6 Batch  900/1077 - Train Accuracy: 0.7223, Validati

Epoch   8 Batch  340/1077 - Train Accuracy: 0.7833, Validation Accuracy: 0.7337, Loss: 0.2245
Epoch   8 Batch  360/1077 - Train Accuracy: 0.7730, Validation Accuracy: 0.7170, Loss: 0.2147
Epoch   8 Batch  380/1077 - Train Accuracy: 0.7539, Validation Accuracy: 0.7305, Loss: 0.2133
Epoch   8 Batch  400/1077 - Train Accuracy: 0.7273, Validation Accuracy: 0.7315, Loss: 0.2456
Epoch   8 Batch  420/1077 - Train Accuracy: 0.7812, Validation Accuracy: 0.7464, Loss: 0.2107
Epoch   8 Batch  440/1077 - Train Accuracy: 0.7660, Validation Accuracy: 0.7280, Loss: 0.2453
Epoch   8 Batch  460/1077 - Train Accuracy: 0.7824, Validation Accuracy: 0.7404, Loss: 0.2348
Epoch   8 Batch  480/1077 - Train Accuracy: 0.7747, Validation Accuracy: 0.7408, Loss: 0.2250
Epoch   8 Batch  500/1077 - Train Accuracy: 0.7387, Validation Accuracy: 0.7347, Loss: 0.2153
Epoch   8 Batch  520/1077 - Train Accuracy: 0.7999, Validation Accuracy: 0.7571, Loss: 0.2045
Epoch   8 Batch  540/1077 - Train Accuracy: 0.7652, Validati

Epoch   9 Batch 1040/1077 - Train Accuracy: 0.8207, Validation Accuracy: 0.7891, Loss: 0.1696
Epoch   9 Batch 1060/1077 - Train Accuracy: 0.8137, Validation Accuracy: 0.7912, Loss: 0.1439
Epoch  10 Batch   20/1077 - Train Accuracy: 0.8199, Validation Accuracy: 0.8161, Loss: 0.1552
Epoch  10 Batch   40/1077 - Train Accuracy: 0.8297, Validation Accuracy: 0.7926, Loss: 0.1534
Epoch  10 Batch   60/1077 - Train Accuracy: 0.8207, Validation Accuracy: 0.7947, Loss: 0.1537
Epoch  10 Batch   80/1077 - Train Accuracy: 0.8535, Validation Accuracy: 0.7955, Loss: 0.1579
Epoch  10 Batch  100/1077 - Train Accuracy: 0.8539, Validation Accuracy: 0.8043, Loss: 0.1576
Epoch  10 Batch  120/1077 - Train Accuracy: 0.8645, Validation Accuracy: 0.8107, Loss: 0.1694
Epoch  10 Batch  140/1077 - Train Accuracy: 0.8664, Validation Accuracy: 0.7887, Loss: 0.1488
Epoch  10 Batch  160/1077 - Train Accuracy: 0.8344, Validation Accuracy: 0.8018, Loss: 0.1528
Epoch  10 Batch  180/1077 - Train Accuracy: 0.8438, Validati

Epoch  11 Batch  680/1077 - Train Accuracy: 0.8348, Validation Accuracy: 0.8228, Loss: 0.1297
Epoch  11 Batch  700/1077 - Train Accuracy: 0.8754, Validation Accuracy: 0.8189, Loss: 0.1028
Epoch  11 Batch  720/1077 - Train Accuracy: 0.8573, Validation Accuracy: 0.8438, Loss: 0.1376
Epoch  11 Batch  740/1077 - Train Accuracy: 0.8867, Validation Accuracy: 0.8366, Loss: 0.1033
Epoch  11 Batch  760/1077 - Train Accuracy: 0.8875, Validation Accuracy: 0.8256, Loss: 0.1235
Epoch  11 Batch  780/1077 - Train Accuracy: 0.8445, Validation Accuracy: 0.8356, Loss: 0.1293
Epoch  11 Batch  800/1077 - Train Accuracy: 0.8871, Validation Accuracy: 0.8612, Loss: 0.1089
Epoch  11 Batch  820/1077 - Train Accuracy: 0.8301, Validation Accuracy: 0.8494, Loss: 0.1266
Epoch  11 Batch  840/1077 - Train Accuracy: 0.8746, Validation Accuracy: 0.8292, Loss: 0.1118
Epoch  11 Batch  860/1077 - Train Accuracy: 0.8415, Validation Accuracy: 0.8494, Loss: 0.1255
Epoch  11 Batch  880/1077 - Train Accuracy: 0.9035, Validati

Epoch  13 Batch  320/1077 - Train Accuracy: 0.9035, Validation Accuracy: 0.8473, Loss: 0.0914
Epoch  13 Batch  340/1077 - Train Accuracy: 0.9289, Validation Accuracy: 0.8331, Loss: 0.0794
Epoch  13 Batch  360/1077 - Train Accuracy: 0.9094, Validation Accuracy: 0.8551, Loss: 0.0796
Epoch  13 Batch  380/1077 - Train Accuracy: 0.9070, Validation Accuracy: 0.8438, Loss: 0.0756
Epoch  13 Batch  400/1077 - Train Accuracy: 0.8723, Validation Accuracy: 0.8288, Loss: 0.1019
Epoch  13 Batch  420/1077 - Train Accuracy: 0.9340, Validation Accuracy: 0.8526, Loss: 0.0772
Epoch  13 Batch  440/1077 - Train Accuracy: 0.8547, Validation Accuracy: 0.8338, Loss: 0.1097
Epoch  13 Batch  460/1077 - Train Accuracy: 0.9137, Validation Accuracy: 0.8271, Loss: 0.0941
Epoch  13 Batch  480/1077 - Train Accuracy: 0.8701, Validation Accuracy: 0.8285, Loss: 0.0888
Epoch  13 Batch  500/1077 - Train Accuracy: 0.9082, Validation Accuracy: 0.8377, Loss: 0.0726
Epoch  13 Batch  520/1077 - Train Accuracy: 0.9193, Validati

Epoch  14 Batch 1020/1077 - Train Accuracy: 0.9023, Validation Accuracy: 0.8384, Loss: 0.0627
Epoch  14 Batch 1040/1077 - Train Accuracy: 0.9219, Validation Accuracy: 0.8750, Loss: 0.0706
Epoch  14 Batch 1060/1077 - Train Accuracy: 0.9133, Validation Accuracy: 0.8523, Loss: 0.0651
Epoch  15 Batch   20/1077 - Train Accuracy: 0.8824, Validation Accuracy: 0.8406, Loss: 0.0629
Epoch  15 Batch   40/1077 - Train Accuracy: 0.9297, Validation Accuracy: 0.8398, Loss: 0.0636
Epoch  15 Batch   60/1077 - Train Accuracy: 0.9051, Validation Accuracy: 0.8512, Loss: 0.0613
Epoch  15 Batch   80/1077 - Train Accuracy: 0.9363, Validation Accuracy: 0.8757, Loss: 0.0673
Epoch  15 Batch  100/1077 - Train Accuracy: 0.9043, Validation Accuracy: 0.8654, Loss: 0.0625
Epoch  15 Batch  120/1077 - Train Accuracy: 0.9223, Validation Accuracy: 0.8612, Loss: 0.0762
Epoch  15 Batch  140/1077 - Train Accuracy: 0.9346, Validation Accuracy: 0.8477, Loss: 0.0578
Epoch  15 Batch  160/1077 - Train Accuracy: 0.8973, Validati

Epoch  16 Batch  660/1077 - Train Accuracy: 0.9309, Validation Accuracy: 0.8661, Loss: 0.0541
Epoch  16 Batch  680/1077 - Train Accuracy: 0.8806, Validation Accuracy: 0.8608, Loss: 0.0565
Epoch  16 Batch  700/1077 - Train Accuracy: 0.9164, Validation Accuracy: 0.8615, Loss: 0.0456
Epoch  16 Batch  720/1077 - Train Accuracy: 0.9248, Validation Accuracy: 0.8697, Loss: 0.0605
Epoch  16 Batch  740/1077 - Train Accuracy: 0.9156, Validation Accuracy: 0.8672, Loss: 0.0494
Epoch  16 Batch  760/1077 - Train Accuracy: 0.9113, Validation Accuracy: 0.8814, Loss: 0.0593
Epoch  16 Batch  780/1077 - Train Accuracy: 0.8875, Validation Accuracy: 0.8714, Loss: 0.0687
Epoch  16 Batch  800/1077 - Train Accuracy: 0.9258, Validation Accuracy: 0.8906, Loss: 0.0517
Epoch  16 Batch  820/1077 - Train Accuracy: 0.8859, Validation Accuracy: 0.8540, Loss: 0.0626
Epoch  16 Batch  840/1077 - Train Accuracy: 0.9051, Validation Accuracy: 0.8707, Loss: 0.0540
Epoch  16 Batch  860/1077 - Train Accuracy: 0.9040, Validati

Epoch  18 Batch  300/1077 - Train Accuracy: 0.9334, Validation Accuracy: 0.9034, Loss: 0.0407
Epoch  18 Batch  320/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.8906, Loss: 0.0482
Epoch  18 Batch  340/1077 - Train Accuracy: 0.9511, Validation Accuracy: 0.8832, Loss: 0.0412
Epoch  18 Batch  360/1077 - Train Accuracy: 0.9410, Validation Accuracy: 0.8725, Loss: 0.0390
Epoch  18 Batch  380/1077 - Train Accuracy: 0.9492, Validation Accuracy: 0.8789, Loss: 0.0384
Epoch  18 Batch  400/1077 - Train Accuracy: 0.9160, Validation Accuracy: 0.8842, Loss: 0.0574
Epoch  18 Batch  420/1077 - Train Accuracy: 0.9574, Validation Accuracy: 0.8828, Loss: 0.0365
Epoch  18 Batch  440/1077 - Train Accuracy: 0.9105, Validation Accuracy: 0.8906, Loss: 0.0549
Epoch  18 Batch  460/1077 - Train Accuracy: 0.9102, Validation Accuracy: 0.8960, Loss: 0.0574
Epoch  18 Batch  480/1077 - Train Accuracy: 0.9235, Validation Accuracy: 0.8860, Loss: 0.0460
Epoch  18 Batch  500/1077 - Train Accuracy: 0.9359, Validati

Epoch  19 Batch 1000/1077 - Train Accuracy: 0.9408, Validation Accuracy: 0.9002, Loss: 0.0410
Epoch  19 Batch 1020/1077 - Train Accuracy: 0.9547, Validation Accuracy: 0.8974, Loss: 0.0330
Epoch  19 Batch 1040/1077 - Train Accuracy: 0.9387, Validation Accuracy: 0.8981, Loss: 0.0416
Epoch  19 Batch 1060/1077 - Train Accuracy: 0.9320, Validation Accuracy: 0.8977, Loss: 0.0363
Epoch  20 Batch   20/1077 - Train Accuracy: 0.9398, Validation Accuracy: 0.9155, Loss: 0.0359
Epoch  20 Batch   40/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9020, Loss: 0.0383
Epoch  20 Batch   60/1077 - Train Accuracy: 0.9241, Validation Accuracy: 0.9130, Loss: 0.0336
Epoch  20 Batch   80/1077 - Train Accuracy: 0.9504, Validation Accuracy: 0.8913, Loss: 0.0419
Epoch  20 Batch  100/1077 - Train Accuracy: 0.9418, Validation Accuracy: 0.8931, Loss: 0.0428
Epoch  20 Batch  120/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9066, Loss: 0.0494
Epoch  20 Batch  140/1077 - Train Accuracy: 0.9502, Validati

Epoch  21 Batch  640/1077 - Train Accuracy: 0.9226, Validation Accuracy: 0.9155, Loss: 0.0320
Epoch  21 Batch  660/1077 - Train Accuracy: 0.9555, Validation Accuracy: 0.8977, Loss: 0.0335
Epoch  21 Batch  680/1077 - Train Accuracy: 0.9260, Validation Accuracy: 0.9190, Loss: 0.0389
Epoch  21 Batch  700/1077 - Train Accuracy: 0.9281, Validation Accuracy: 0.9077, Loss: 0.0335
Epoch  21 Batch  720/1077 - Train Accuracy: 0.9243, Validation Accuracy: 0.9098, Loss: 0.0400
Epoch  21 Batch  740/1077 - Train Accuracy: 0.9375, Validation Accuracy: 0.9219, Loss: 0.0326
Epoch  21 Batch  760/1077 - Train Accuracy: 0.9359, Validation Accuracy: 0.9116, Loss: 0.0402
Epoch  21 Batch  780/1077 - Train Accuracy: 0.9156, Validation Accuracy: 0.9077, Loss: 0.0479
Epoch  21 Batch  800/1077 - Train Accuracy: 0.9465, Validation Accuracy: 0.9162, Loss: 0.0318
Epoch  21 Batch  820/1077 - Train Accuracy: 0.9461, Validation Accuracy: 0.9006, Loss: 0.0399
Epoch  21 Batch  840/1077 - Train Accuracy: 0.9676, Validati

Epoch  23 Batch  280/1077 - Train Accuracy: 0.9016, Validation Accuracy: 0.9027, Loss: 0.0393
Epoch  23 Batch  300/1077 - Train Accuracy: 0.9618, Validation Accuracy: 0.9066, Loss: 0.0274
Epoch  23 Batch  320/1077 - Train Accuracy: 0.9465, Validation Accuracy: 0.9197, Loss: 0.0367
Epoch  23 Batch  340/1077 - Train Accuracy: 0.9498, Validation Accuracy: 0.9176, Loss: 0.0320
Epoch  23 Batch  360/1077 - Train Accuracy: 0.9426, Validation Accuracy: 0.9116, Loss: 0.0262
Epoch  23 Batch  380/1077 - Train Accuracy: 0.9469, Validation Accuracy: 0.9180, Loss: 0.0265
Epoch  23 Batch  400/1077 - Train Accuracy: 0.9437, Validation Accuracy: 0.9130, Loss: 0.0395
Epoch  23 Batch  420/1077 - Train Accuracy: 0.9730, Validation Accuracy: 0.9208, Loss: 0.0249
Epoch  23 Batch  440/1077 - Train Accuracy: 0.9156, Validation Accuracy: 0.9194, Loss: 0.0399
Epoch  23 Batch  460/1077 - Train Accuracy: 0.9234, Validation Accuracy: 0.9183, Loss: 0.0394
Epoch  23 Batch  480/1077 - Train Accuracy: 0.9305, Validati

Epoch  24 Batch  980/1077 - Train Accuracy: 0.9340, Validation Accuracy: 0.9400, Loss: 0.0341
Epoch  24 Batch 1000/1077 - Train Accuracy: 0.9304, Validation Accuracy: 0.9226, Loss: 0.0306
Epoch  24 Batch 1020/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9201, Loss: 0.0220
Epoch  24 Batch 1040/1077 - Train Accuracy: 0.9502, Validation Accuracy: 0.9286, Loss: 0.0311
Epoch  24 Batch 1060/1077 - Train Accuracy: 0.9582, Validation Accuracy: 0.9343, Loss: 0.0266
Epoch  25 Batch   20/1077 - Train Accuracy: 0.9551, Validation Accuracy: 0.9308, Loss: 0.0272
Epoch  25 Batch   40/1077 - Train Accuracy: 0.9453, Validation Accuracy: 0.9396, Loss: 0.0261
Epoch  25 Batch   60/1077 - Train Accuracy: 0.9524, Validation Accuracy: 0.9350, Loss: 0.0234
Epoch  25 Batch   80/1077 - Train Accuracy: 0.9492, Validation Accuracy: 0.9162, Loss: 0.0302
Epoch  25 Batch  100/1077 - Train Accuracy: 0.9473, Validation Accuracy: 0.9226, Loss: 0.0341
Epoch  25 Batch  120/1077 - Train Accuracy: 0.9379, Validati

Epoch  26 Batch  620/1077 - Train Accuracy: 0.9668, Validation Accuracy: 0.9442, Loss: 0.0270
Epoch  26 Batch  640/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9389, Loss: 0.0254
Epoch  26 Batch  660/1077 - Train Accuracy: 0.9676, Validation Accuracy: 0.9297, Loss: 0.0258
Epoch  26 Batch  680/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9300, Loss: 0.0318
Epoch  26 Batch  700/1077 - Train Accuracy: 0.9559, Validation Accuracy: 0.9258, Loss: 0.0245
Epoch  26 Batch  720/1077 - Train Accuracy: 0.9359, Validation Accuracy: 0.9343, Loss: 0.0316
Epoch  26 Batch  740/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9268, Loss: 0.0269
Epoch  26 Batch  760/1077 - Train Accuracy: 0.9453, Validation Accuracy: 0.9393, Loss: 0.0306
Epoch  26 Batch  780/1077 - Train Accuracy: 0.9348, Validation Accuracy: 0.9371, Loss: 0.0418
Epoch  26 Batch  800/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9361, Loss: 0.0258
Epoch  26 Batch  820/1077 - Train Accuracy: 0.9605, Validati

Epoch  28 Batch  260/1077 - Train Accuracy: 0.9167, Validation Accuracy: 0.9279, Loss: 0.0233
Epoch  28 Batch  280/1077 - Train Accuracy: 0.9234, Validation Accuracy: 0.9311, Loss: 0.0334
Epoch  28 Batch  300/1077 - Train Accuracy: 0.9655, Validation Accuracy: 0.9205, Loss: 0.0217
Epoch  28 Batch  320/1077 - Train Accuracy: 0.9453, Validation Accuracy: 0.9311, Loss: 0.0289
Epoch  28 Batch  340/1077 - Train Accuracy: 0.9589, Validation Accuracy: 0.9318, Loss: 0.0275
Epoch  28 Batch  360/1077 - Train Accuracy: 0.9699, Validation Accuracy: 0.9350, Loss: 0.0204
Epoch  28 Batch  380/1077 - Train Accuracy: 0.9641, Validation Accuracy: 0.9247, Loss: 0.0228
Epoch  28 Batch  400/1077 - Train Accuracy: 0.9336, Validation Accuracy: 0.9290, Loss: 0.0283
Epoch  28 Batch  420/1077 - Train Accuracy: 0.9836, Validation Accuracy: 0.9244, Loss: 0.0187
Epoch  28 Batch  440/1077 - Train Accuracy: 0.9359, Validation Accuracy: 0.9261, Loss: 0.0312
Epoch  28 Batch  460/1077 - Train Accuracy: 0.9352, Validati

Epoch  29 Batch  960/1077 - Train Accuracy: 0.9464, Validation Accuracy: 0.9382, Loss: 0.0214
Epoch  29 Batch  980/1077 - Train Accuracy: 0.9297, Validation Accuracy: 0.9489, Loss: 0.0275
Epoch  29 Batch 1000/1077 - Train Accuracy: 0.9334, Validation Accuracy: 0.9336, Loss: 0.0253
Epoch  29 Batch 1020/1077 - Train Accuracy: 0.9594, Validation Accuracy: 0.9382, Loss: 0.0183
Epoch  29 Batch 1040/1077 - Train Accuracy: 0.9700, Validation Accuracy: 0.9297, Loss: 0.0243
Epoch  29 Batch 1060/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9563, Loss: 0.0211
Epoch  30 Batch   20/1077 - Train Accuracy: 0.9660, Validation Accuracy: 0.9478, Loss: 0.0191
Epoch  30 Batch   40/1077 - Train Accuracy: 0.9590, Validation Accuracy: 0.9535, Loss: 0.0216
Epoch  30 Batch   60/1077 - Train Accuracy: 0.9714, Validation Accuracy: 0.9279, Loss: 0.0177
Epoch  30 Batch   80/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9421, Loss: 0.0242
Epoch  30 Batch  100/1077 - Train Accuracy: 0.9488, Validati

Epoch  31 Batch  600/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9467, Loss: 0.0260
Epoch  31 Batch  620/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9513, Loss: 0.0218
Epoch  31 Batch  640/1077 - Train Accuracy: 0.9665, Validation Accuracy: 0.9354, Loss: 0.0194
Epoch  31 Batch  660/1077 - Train Accuracy: 0.9762, Validation Accuracy: 0.9432, Loss: 0.0202
Epoch  31 Batch  680/1077 - Train Accuracy: 0.9442, Validation Accuracy: 0.9411, Loss: 0.0271
Epoch  31 Batch  700/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9343, Loss: 0.0195
Epoch  31 Batch  720/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9471, Loss: 0.0253
Epoch  31 Batch  740/1077 - Train Accuracy: 0.9664, Validation Accuracy: 0.9467, Loss: 0.0218
Epoch  31 Batch  760/1077 - Train Accuracy: 0.9516, Validation Accuracy: 0.9492, Loss: 0.0222
Epoch  31 Batch  780/1077 - Train Accuracy: 0.9434, Validation Accuracy: 0.9432, Loss: 0.0340
Epoch  31 Batch  800/1077 - Train Accuracy: 0.9633, Validati

Epoch  33 Batch  240/1077 - Train Accuracy: 0.9648, Validation Accuracy: 0.9574, Loss: 0.0172
Epoch  33 Batch  260/1077 - Train Accuracy: 0.9304, Validation Accuracy: 0.9485, Loss: 0.0203
Epoch  33 Batch  280/1077 - Train Accuracy: 0.9348, Validation Accuracy: 0.9513, Loss: 0.0268
Epoch  33 Batch  300/1077 - Train Accuracy: 0.9650, Validation Accuracy: 0.9364, Loss: 0.0202
Epoch  33 Batch  320/1077 - Train Accuracy: 0.9453, Validation Accuracy: 0.9613, Loss: 0.0233
Epoch  33 Batch  340/1077 - Train Accuracy: 0.9597, Validation Accuracy: 0.9439, Loss: 0.0215
Epoch  33 Batch  360/1077 - Train Accuracy: 0.9734, Validation Accuracy: 0.9393, Loss: 0.0149
Epoch  33 Batch  380/1077 - Train Accuracy: 0.9637, Validation Accuracy: 0.9545, Loss: 0.0206
Epoch  33 Batch  400/1077 - Train Accuracy: 0.9563, Validation Accuracy: 0.9485, Loss: 0.0220
Epoch  33 Batch  420/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9428, Loss: 0.0147
Epoch  33 Batch  440/1077 - Train Accuracy: 0.9375, Validati

Epoch  34 Batch  940/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9666, Loss: 0.0197
Epoch  34 Batch  960/1077 - Train Accuracy: 0.9528, Validation Accuracy: 0.9503, Loss: 0.0176
Epoch  34 Batch  980/1077 - Train Accuracy: 0.9516, Validation Accuracy: 0.9517, Loss: 0.0231
Epoch  34 Batch 1000/1077 - Train Accuracy: 0.9405, Validation Accuracy: 0.9375, Loss: 0.0220
Epoch  34 Batch 1020/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9645, Loss: 0.0142
Epoch  34 Batch 1040/1077 - Train Accuracy: 0.9720, Validation Accuracy: 0.9641, Loss: 0.0195
Epoch  34 Batch 1060/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9712, Loss: 0.0175
Epoch  35 Batch   20/1077 - Train Accuracy: 0.9828, Validation Accuracy: 0.9517, Loss: 0.0134
Epoch  35 Batch   40/1077 - Train Accuracy: 0.9598, Validation Accuracy: 0.9712, Loss: 0.0168
Epoch  35 Batch   60/1077 - Train Accuracy: 0.9658, Validation Accuracy: 0.9556, Loss: 0.0155
Epoch  35 Batch   80/1077 - Train Accuracy: 0.9656, Validati

Epoch  36 Batch  580/1077 - Train Accuracy: 0.9717, Validation Accuracy: 0.9570, Loss: 0.0146
Epoch  36 Batch  600/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9570, Loss: 0.0239
Epoch  36 Batch  620/1077 - Train Accuracy: 0.9789, Validation Accuracy: 0.9592, Loss: 0.0177
Epoch  36 Batch  640/1077 - Train Accuracy: 0.9736, Validation Accuracy: 0.9567, Loss: 0.0159
Epoch  36 Batch  660/1077 - Train Accuracy: 0.9824, Validation Accuracy: 0.9545, Loss: 0.0162
Epoch  36 Batch  680/1077 - Train Accuracy: 0.9449, Validation Accuracy: 0.9474, Loss: 0.0257
Epoch  36 Batch  700/1077 - Train Accuracy: 0.9828, Validation Accuracy: 0.9545, Loss: 0.0174
Epoch  36 Batch  720/1077 - Train Accuracy: 0.9535, Validation Accuracy: 0.9606, Loss: 0.0242
Epoch  36 Batch  740/1077 - Train Accuracy: 0.9586, Validation Accuracy: 0.9592, Loss: 0.0177
Epoch  36 Batch  760/1077 - Train Accuracy: 0.9625, Validation Accuracy: 0.9624, Loss: 0.0182
Epoch  36 Batch  780/1077 - Train Accuracy: 0.9473, Validati

Epoch  38 Batch  220/1077 - Train Accuracy: 0.9560, Validation Accuracy: 0.9627, Loss: 0.0285
Epoch  38 Batch  240/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9684, Loss: 0.0135
Epoch  38 Batch  260/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9492, Loss: 0.0167
Epoch  38 Batch  280/1077 - Train Accuracy: 0.9469, Validation Accuracy: 0.9542, Loss: 0.0239
Epoch  38 Batch  300/1077 - Train Accuracy: 0.9683, Validation Accuracy: 0.9432, Loss: 0.0184
Epoch  38 Batch  320/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9627, Loss: 0.0203
Epoch  38 Batch  340/1077 - Train Accuracy: 0.9782, Validation Accuracy: 0.9638, Loss: 0.0181
Epoch  38 Batch  360/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9517, Loss: 0.0145
Epoch  38 Batch  380/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9620, Loss: 0.0161
Epoch  38 Batch  400/1077 - Train Accuracy: 0.9563, Validation Accuracy: 0.9549, Loss: 0.0198
Epoch  38 Batch  420/1077 - Train Accuracy: 0.9852, Validati

Epoch  39 Batch  920/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9588, Loss: 0.0160
Epoch  39 Batch  940/1077 - Train Accuracy: 0.9566, Validation Accuracy: 0.9673, Loss: 0.0173
Epoch  39 Batch  960/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9506, Loss: 0.0150
Epoch  39 Batch  980/1077 - Train Accuracy: 0.9488, Validation Accuracy: 0.9499, Loss: 0.0201
Epoch  39 Batch 1000/1077 - Train Accuracy: 0.9468, Validation Accuracy: 0.9560, Loss: 0.0187
Epoch  39 Batch 1020/1077 - Train Accuracy: 0.9707, Validation Accuracy: 0.9602, Loss: 0.0116
Epoch  39 Batch 1040/1077 - Train Accuracy: 0.9737, Validation Accuracy: 0.9648, Loss: 0.0164
Epoch  39 Batch 1060/1077 - Train Accuracy: 0.9938, Validation Accuracy: 0.9723, Loss: 0.0140
Epoch  40 Batch   20/1077 - Train Accuracy: 0.9828, Validation Accuracy: 0.9638, Loss: 0.0112
Epoch  40 Batch   40/1077 - Train Accuracy: 0.9563, Validation Accuracy: 0.9688, Loss: 0.0155
Epoch  40 Batch   60/1077 - Train Accuracy: 0.9721, Validati

Epoch  41 Batch  560/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9521, Loss: 0.0132
Epoch  41 Batch  580/1077 - Train Accuracy: 0.9795, Validation Accuracy: 0.9719, Loss: 0.0150
Epoch  41 Batch  600/1077 - Train Accuracy: 0.9814, Validation Accuracy: 0.9620, Loss: 0.0192
Epoch  41 Batch  620/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9620, Loss: 0.0168
Epoch  41 Batch  640/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9616, Loss: 0.0134
Epoch  41 Batch  660/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9545, Loss: 0.0147
Epoch  41 Batch  680/1077 - Train Accuracy: 0.9494, Validation Accuracy: 0.9560, Loss: 0.0247
Epoch  41 Batch  700/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9581, Loss: 0.0156
Epoch  41 Batch  720/1077 - Train Accuracy: 0.9576, Validation Accuracy: 0.9684, Loss: 0.0200
Epoch  41 Batch  740/1077 - Train Accuracy: 0.9727, Validation Accuracy: 0.9666, Loss: 0.0147
Epoch  41 Batch  760/1077 - Train Accuracy: 0.9699, Validati

Epoch  43 Batch  200/1077 - Train Accuracy: 0.9773, Validation Accuracy: 0.9737, Loss: 0.0142
Epoch  43 Batch  220/1077 - Train Accuracy: 0.9659, Validation Accuracy: 0.9680, Loss: 0.0224
Epoch  43 Batch  240/1077 - Train Accuracy: 0.9852, Validation Accuracy: 0.9652, Loss: 0.0111
Epoch  43 Batch  260/1077 - Train Accuracy: 0.9308, Validation Accuracy: 0.9620, Loss: 0.0150
Epoch  43 Batch  280/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9489, Loss: 0.0204
Epoch  43 Batch  300/1077 - Train Accuracy: 0.9696, Validation Accuracy: 0.9624, Loss: 0.0170
Epoch  43 Batch  320/1077 - Train Accuracy: 0.9730, Validation Accuracy: 0.9581, Loss: 0.0178
Epoch  43 Batch  340/1077 - Train Accuracy: 0.9807, Validation Accuracy: 0.9620, Loss: 0.0170
Epoch  43 Batch  360/1077 - Train Accuracy: 0.9840, Validation Accuracy: 0.9503, Loss: 0.0109
Epoch  43 Batch  380/1077 - Train Accuracy: 0.9711, Validation Accuracy: 0.9585, Loss: 0.0124
Epoch  43 Batch  400/1077 - Train Accuracy: 0.9707, Validati

Epoch  44 Batch  900/1077 - Train Accuracy: 0.9805, Validation Accuracy: 0.9666, Loss: 0.0176
Epoch  44 Batch  920/1077 - Train Accuracy: 0.9785, Validation Accuracy: 0.9613, Loss: 0.0141
Epoch  44 Batch  940/1077 - Train Accuracy: 0.9625, Validation Accuracy: 0.9574, Loss: 0.0139
Epoch  44 Batch  960/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9599, Loss: 0.0125
Epoch  44 Batch  980/1077 - Train Accuracy: 0.9488, Validation Accuracy: 0.9599, Loss: 0.0161
Epoch  44 Batch 1000/1077 - Train Accuracy: 0.9509, Validation Accuracy: 0.9616, Loss: 0.0153
Epoch  44 Batch 1020/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9716, Loss: 0.0105
Epoch  44 Batch 1040/1077 - Train Accuracy: 0.9737, Validation Accuracy: 0.9648, Loss: 0.0152
Epoch  44 Batch 1060/1077 - Train Accuracy: 0.9980, Validation Accuracy: 0.9545, Loss: 0.0114
Epoch  45 Batch   20/1077 - Train Accuracy: 0.9816, Validation Accuracy: 0.9645, Loss: 0.0111
Epoch  45 Batch   40/1077 - Train Accuracy: 0.9738, Validati

Epoch  46 Batch  540/1077 - Train Accuracy: 0.9672, Validation Accuracy: 0.9602, Loss: 0.0132
Epoch  46 Batch  560/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9560, Loss: 0.0125
Epoch  46 Batch  580/1077 - Train Accuracy: 0.9792, Validation Accuracy: 0.9620, Loss: 0.0126
Epoch  46 Batch  600/1077 - Train Accuracy: 0.9751, Validation Accuracy: 0.9585, Loss: 0.0176
Epoch  46 Batch  620/1077 - Train Accuracy: 0.9816, Validation Accuracy: 0.9670, Loss: 0.0142
Epoch  46 Batch  640/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9659, Loss: 0.0135
Epoch  46 Batch  660/1077 - Train Accuracy: 0.9867, Validation Accuracy: 0.9585, Loss: 0.0112
Epoch  46 Batch  680/1077 - Train Accuracy: 0.9457, Validation Accuracy: 0.9577, Loss: 0.0221
Epoch  46 Batch  700/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9496, Loss: 0.0134
Epoch  46 Batch  720/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9656, Loss: 0.0189
Epoch  46 Batch  740/1077 - Train Accuracy: 0.9801, Validati

Epoch  48 Batch  180/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9513, Loss: 0.0109
Epoch  48 Batch  200/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9727, Loss: 0.0123
Epoch  48 Batch  220/1077 - Train Accuracy: 0.9741, Validation Accuracy: 0.9691, Loss: 0.0180
Epoch  48 Batch  240/1077 - Train Accuracy: 0.9902, Validation Accuracy: 0.9656, Loss: 0.0093
Epoch  48 Batch  260/1077 - Train Accuracy: 0.9628, Validation Accuracy: 0.9592, Loss: 0.0134
Epoch  48 Batch  280/1077 - Train Accuracy: 0.9617, Validation Accuracy: 0.9638, Loss: 0.0181
Epoch  48 Batch  300/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9673, Loss: 0.0128
Epoch  48 Batch  320/1077 - Train Accuracy: 0.9727, Validation Accuracy: 0.9641, Loss: 0.0160
Epoch  48 Batch  340/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9709, Loss: 0.0153
Epoch  48 Batch  360/1077 - Train Accuracy: 0.9742, Validation Accuracy: 0.9549, Loss: 0.0089
Epoch  48 Batch  380/1077 - Train Accuracy: 0.9828, Validati

Epoch  49 Batch  880/1077 - Train Accuracy: 0.9828, Validation Accuracy: 0.9659, Loss: 0.0148
Epoch  49 Batch  900/1077 - Train Accuracy: 0.9824, Validation Accuracy: 0.9730, Loss: 0.0171
Epoch  49 Batch  920/1077 - Train Accuracy: 0.9836, Validation Accuracy: 0.9585, Loss: 0.0128
Epoch  49 Batch  940/1077 - Train Accuracy: 0.9746, Validation Accuracy: 0.9787, Loss: 0.0117
Epoch  49 Batch  960/1077 - Train Accuracy: 0.9669, Validation Accuracy: 0.9592, Loss: 0.0130
Epoch  49 Batch  980/1077 - Train Accuracy: 0.9660, Validation Accuracy: 0.9613, Loss: 0.0129
Epoch  49 Batch 1000/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9702, Loss: 0.0146
Epoch  49 Batch 1020/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9581, Loss: 0.0094
Epoch  49 Batch 1040/1077 - Train Accuracy: 0.9786, Validation Accuracy: 0.9616, Loss: 0.0142
Epoch  49 Batch 1060/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9691, Loss: 0.0117
Epoch  50 Batch   20/1077 - Train Accuracy: 0.9832, Validati

Epoch  51 Batch  520/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9570, Loss: 0.0072
Epoch  51 Batch  540/1077 - Train Accuracy: 0.9711, Validation Accuracy: 0.9616, Loss: 0.0116
Epoch  51 Batch  560/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9624, Loss: 0.0119
Epoch  51 Batch  580/1077 - Train Accuracy: 0.9795, Validation Accuracy: 0.9641, Loss: 0.0095
Epoch  51 Batch  600/1077 - Train Accuracy: 0.9825, Validation Accuracy: 0.9482, Loss: 0.0182
Epoch  51 Batch  620/1077 - Train Accuracy: 0.9832, Validation Accuracy: 0.9727, Loss: 0.0141
Epoch  51 Batch  640/1077 - Train Accuracy: 0.9769, Validation Accuracy: 0.9620, Loss: 0.0110
Epoch  51 Batch  660/1077 - Train Accuracy: 0.9918, Validation Accuracy: 0.9592, Loss: 0.0100
Epoch  51 Batch  680/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9478, Loss: 0.0205
Epoch  51 Batch  700/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9503, Loss: 0.0114
Epoch  51 Batch  720/1077 - Train Accuracy: 0.9593, Validati

Epoch  53 Batch  160/1077 - Train Accuracy: 0.9809, Validation Accuracy: 0.9688, Loss: 0.0134
Epoch  53 Batch  180/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9517, Loss: 0.0105
Epoch  53 Batch  200/1077 - Train Accuracy: 0.9844, Validation Accuracy: 0.9766, Loss: 0.0104
Epoch  53 Batch  220/1077 - Train Accuracy: 0.9741, Validation Accuracy: 0.9680, Loss: 0.0151
Epoch  53 Batch  240/1077 - Train Accuracy: 0.9898, Validation Accuracy: 0.9616, Loss: 0.0075
Epoch  53 Batch  260/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9624, Loss: 0.0111
Epoch  53 Batch  280/1077 - Train Accuracy: 0.9559, Validation Accuracy: 0.9663, Loss: 0.0167
Epoch  53 Batch  300/1077 - Train Accuracy: 0.9745, Validation Accuracy: 0.9719, Loss: 0.0107
Epoch  53 Batch  320/1077 - Train Accuracy: 0.9809, Validation Accuracy: 0.9631, Loss: 0.0153
Epoch  53 Batch  340/1077 - Train Accuracy: 0.9893, Validation Accuracy: 0.9670, Loss: 0.0139
Epoch  53 Batch  360/1077 - Train Accuracy: 0.9816, Validati

Epoch  54 Batch  860/1077 - Train Accuracy: 0.9743, Validation Accuracy: 0.9702, Loss: 0.0163
Epoch  54 Batch  880/1077 - Train Accuracy: 0.9832, Validation Accuracy: 0.9712, Loss: 0.0129
Epoch  54 Batch  900/1077 - Train Accuracy: 0.9805, Validation Accuracy: 0.9592, Loss: 0.0161
Epoch  54 Batch  920/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9627, Loss: 0.0119
Epoch  54 Batch  940/1077 - Train Accuracy: 0.9746, Validation Accuracy: 0.9716, Loss: 0.0130
Epoch  54 Batch  960/1077 - Train Accuracy: 0.9676, Validation Accuracy: 0.9670, Loss: 0.0111
Epoch  54 Batch  980/1077 - Train Accuracy: 0.9715, Validation Accuracy: 0.9673, Loss: 0.0126
Epoch  54 Batch 1000/1077 - Train Accuracy: 0.9773, Validation Accuracy: 0.9663, Loss: 0.0123
Epoch  54 Batch 1020/1077 - Train Accuracy: 0.9934, Validation Accuracy: 0.9677, Loss: 0.0082
Epoch  54 Batch 1040/1077 - Train Accuracy: 0.9778, Validation Accuracy: 0.9620, Loss: 0.0121
Epoch  54 Batch 1060/1077 - Train Accuracy: 0.9879, Validati

Epoch  56 Batch  500/1077 - Train Accuracy: 0.9809, Validation Accuracy: 0.9542, Loss: 0.0088
Epoch  56 Batch  520/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9606, Loss: 0.0066
Epoch  56 Batch  540/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9613, Loss: 0.0095
Epoch  56 Batch  560/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9684, Loss: 0.0097
Epoch  56 Batch  580/1077 - Train Accuracy: 0.9795, Validation Accuracy: 0.9695, Loss: 0.0106
Epoch  56 Batch  600/1077 - Train Accuracy: 0.9825, Validation Accuracy: 0.9638, Loss: 0.0155
Epoch  56 Batch  620/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9719, Loss: 0.0125
Epoch  56 Batch  640/1077 - Train Accuracy: 0.9725, Validation Accuracy: 0.9673, Loss: 0.0110
Epoch  56 Batch  660/1077 - Train Accuracy: 0.9930, Validation Accuracy: 0.9457, Loss: 0.0099
Epoch  56 Batch  680/1077 - Train Accuracy: 0.9628, Validation Accuracy: 0.9524, Loss: 0.0170
Epoch  56 Batch  700/1077 - Train Accuracy: 0.9844, Validati

Epoch  58 Batch  140/1077 - Train Accuracy: 0.9827, Validation Accuracy: 0.9553, Loss: 0.0095
Epoch  58 Batch  160/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9702, Loss: 0.0110
Epoch  58 Batch  180/1077 - Train Accuracy: 0.9809, Validation Accuracy: 0.9545, Loss: 0.0101
Epoch  58 Batch  200/1077 - Train Accuracy: 0.9824, Validation Accuracy: 0.9748, Loss: 0.0094
Epoch  58 Batch  220/1077 - Train Accuracy: 0.9794, Validation Accuracy: 0.9680, Loss: 0.0142
Epoch  58 Batch  240/1077 - Train Accuracy: 0.9941, Validation Accuracy: 0.9624, Loss: 0.0072
Epoch  58 Batch  260/1077 - Train Accuracy: 0.9658, Validation Accuracy: 0.9556, Loss: 0.0088
Epoch  58 Batch  280/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9673, Loss: 0.0156
Epoch  58 Batch  300/1077 - Train Accuracy: 0.9803, Validation Accuracy: 0.9719, Loss: 0.0105
Epoch  58 Batch  320/1077 - Train Accuracy: 0.9719, Validation Accuracy: 0.9531, Loss: 0.0133
Epoch  58 Batch  340/1077 - Train Accuracy: 0.9889, Validati

Epoch  59 Batch  840/1077 - Train Accuracy: 0.9723, Validation Accuracy: 0.9645, Loss: 0.0143
Epoch  59 Batch  860/1077 - Train Accuracy: 0.9769, Validation Accuracy: 0.9624, Loss: 0.0129
Epoch  59 Batch  880/1077 - Train Accuracy: 0.9902, Validation Accuracy: 0.9751, Loss: 0.0099
Epoch  59 Batch  900/1077 - Train Accuracy: 0.9746, Validation Accuracy: 0.9634, Loss: 0.0129
Epoch  59 Batch  920/1077 - Train Accuracy: 0.9852, Validation Accuracy: 0.9670, Loss: 0.0112
Epoch  59 Batch  940/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9698, Loss: 0.0100
Epoch  59 Batch  960/1077 - Train Accuracy: 0.9717, Validation Accuracy: 0.9624, Loss: 0.0091
Epoch  59 Batch  980/1077 - Train Accuracy: 0.9711, Validation Accuracy: 0.9645, Loss: 0.0114
Epoch  59 Batch 1000/1077 - Train Accuracy: 0.9814, Validation Accuracy: 0.9691, Loss: 0.0114
Epoch  59 Batch 1020/1077 - Train Accuracy: 0.9895, Validation Accuracy: 0.9744, Loss: 0.0077
Epoch  59 Batch 1040/1077 - Train Accuracy: 0.9696, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [20]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    '''Prepare the text for the model'''
    sentence = sentence.lower()
    sequence_length = 7
    return [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [21]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [34, 69, 116, 113, 199, 108, 190]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [219, 232, 152, 350, 142, 160, 87, 331, 1]
  French Words: il a vu une voiture à l'école ? <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.